[iOS] Keyboard scrolling does not work in sub-scrollable regions

rr-codes opened the pull request:
#### 3b564766bf831003582c5bc07c0975854a5478e8
<pre>
[iOS] Keyboard scrolling does not work in sub-scrollable regions
<a href="https://bugs.webkit.org/show_bug.cgi?id=287721">https://bugs.webkit.org/show_bug.cgi?id=287721</a>
<a href="https://rdar.apple.com/131354460">rdar://131354460</a>

Reviewed by Tim Horton.

Add functionality to allow scrolling in sub-scrollable regions on iOS like macOS.

* LayoutTests/fast/scrolling/ios/keyboard-scrolling-subscrollable-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/keyboard-scrolling-subscrollable.html: Added.

Layout tests.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::focusedScrollableArea const):

Fallback to the last touched node if there is no mouse press node or focused node.

* Source/WebCore/page/Page.h:
(WebCore::Page::setLastTouchedNode):
(WebCore::Page::lastTouchedNode const):

Store the last touched node so that EventHandler can later use it.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _scrollViewInternal]):
* Source/WebKit/UIProcess/ios/WKVelocityTrackingScrollView.h: Removed.
* Source/WebKit/UIProcess/ios/WKVelocityTrackingScrollView.mm: Removed.

Remove WKVelocityTrackingScrollView in favor of using WKBaseScrollView directly so that all scroll views can track velocities.

* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(ScrollingDeltaWindow::update):
(ScrollingDeltaWindow::reset):
(ScrollingDeltaWindow::averageVelocity const):
(-[WKBaseScrollView updateInteractiveScrollVelocity]):
(-[WKBaseScrollView interactiveScrollVelocityInPointsPerSecond]):

Give WKBaseScrollView the functionality that WKVelocityTrackingScrollView had.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(-[WKScrollingNodeScrollViewDelegate scrollViewDidScroll:]):

For sub-scrollable regions, call `updateInteractiveScrollVelocity` just like the main scroll view.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):

Since the scroll view that the scrolling animator uses is now dynamic, don&apos;t provide a scroll view when creating the animator;
instead, provide it on-demand.

(-[WKContentView _interpretKeyEvent:withContext:]):

Compute the scroll view that should be used for the keyboard scrolling animator, based on whichever view was focused/clicked/touched last.
Fallback to the main scroll view if no others are found.

(-[WKContentView keyboardScrollViewAnimatorDidFinishScrolling:]):

Use the correct scroll view when passing it to `_didFinishScrolling`.

* Source/WebKit/UIProcess/ios/WKKeyboardScrollingAnimator.h:
* Source/WebKit/UIProcess/ios/WKKeyboardScrollingAnimator.mm:
(-[WKKeyboardScrollViewAnimator init]):
(-[WKKeyboardScrollViewAnimator initWithScrollView:]): Deleted.

Instead of providing a scroll view at the beginning, it is provided on-demand.

(-[WKKeyboardScrollViewAnimator beginWithEvent:scrollView:]):
(-[WKKeyboardScrollViewAnimator beginWithEvent:]): Deleted.

Amend this function to allow specifying the current scroll view that should be targeted.
If there is no ongoing scroll, set the ivar _scrollView to the parameter&apos;s value.
If there is an ongoing scroll on a different scroll view, ignore the request on the new scroll view.

(-[WKKeyboardScrollViewAnimator scrollView]):

Allow querying for the current scroll view that the animator is using.

(-[WKKeyboardScrollViewAnimator didFinishScrolling]):

Set _scrollView to nil to indicate that there is now no scroll view that is currently keyboard scrolling.

* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView web_handleKeyEvent:]):
(-[WKPDFView web_initWithFrame:webView:mimeType:]):

Update callsites to use the new functions.

* Source/WebKit/UIProcess/ios/WKScrollView.h:

Change the base class.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Add a new member variable to keep track of the most recently tapped node.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleEditingKeyboardEvent):
(WebKit::WebPage::commitPotentialTap):

If there is no click node, then instead do a hit-test to find the relevant node that can be tapped, and update the last touched node.

Canonical link: <a href="https://commits.webkit.org/292342@main">https://commits.webkit.org/292342@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a00ea1fe8f86f0149ff01d1f31701af35c48af4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95742 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100794 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/46247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97786 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23783 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/73012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/46247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98745 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/11707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/86470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/53346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/11432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4198 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/45586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4308 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102827 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22800 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23052 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82497 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81414 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20383 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/25988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/3446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22768 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/27917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/22427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/25903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/24169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/scrolling/ios/keyboard-scrolling-subscrollable.html:
```diff
@@ -0,0 +1,121 @@
+<html>
+<head>
+    <script src="../../../resources/ui-helper.js"></script>
+    <script src="../../../resources/js-test-pre.js"></script>
```
Use `js-test.js` and remove pre/post.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1369,13 +1369,24 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
+        if (localMainFrame) {
+            constexpr OptionSet hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::AllowVisibleChildFrameContentOnly };
+            auto roundedPoint = IntPoint { m_potentialTapLocation };
+            HitTestResult result = localMainFrame->eventHandler().hitTestResultAtPoint(roundedPoint, hitType);
```
```suggestion
            auto result = localMainFrame->eventHandler().hitTestResultAtPoint(roundedPoint, hitType);
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2835,6 +2835,7 @@ class WebPage final : public API::ObjectImpl<API::Object::Type::BundlePage>, pub
     BidiSelectionFlippingState m_bidiSelectionFlippingState { BidiSelectionFlippingState::NotFlipping };
 
     RefPtr<WebCore::Node> m_potentialTapNode;
+    RefPtr<WebCore::Node> m_lastTouchedNode;
```
This appears unused?

pxlcoder commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -16989,6 +16985,7 @@
 				1C5DC46F290B27260061EC62 /* JSWebExtensionWrappable.h in Headers */,
 				1C5DC46E290B27260061EC62 /* JSWebExtensionWrapper.h in Headers */,
 				074A6FC72D5F1FAF0027F958 /* KeyEventInterpretationContext.h in Headers */,
+				074A6FC72D5F1FAF0027F958 /* KeyEventInterpretationContext.h in Headers */,
```
This is weird, same line appears twice.

pxlcoder commented on Source/WebCore/page/Page.h:
```diff
@@ -1440,6 +1443,8 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     bool m_defersLoading { false };
     unsigned m_defersLoadingCallCount { 0 };
 
+    RefPtr<Node> m_lastTouchedNode;
```
Looking at the existing members on `Page` and `EventHandler`, I think this would make more sense on `EventHandler`.

hortont424 replied:
Agreed

hortont424 reviewed (approved):
agree with aditya's comments, otherwise seems fine

rr-codes committed:
[iOS] Keyboard scrolling does not work in sub-scrollable regions
https://bugs.webkit.org/show_bug.cgi?id=287721
rdar://131354460

Reviewed by Tim Horton.

Add functionality to allow scrolling in sub-scrollable regions on iOS like macOS.

* LayoutTests/fast/scrolling/ios/keyboard-scrolling-subscrollable-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/keyboard-scrolling-subscrollable.html: Added.

Layout tests.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::focusedScrollableArea const):

Fallback to the last touched node if there is no mouse press node or focused node.

* Source/WebCore/page/Page.h:
(WebCore::Page::setLastTouchedNode):
(WebCore::Page::lastTouchedNode const):

Store the last touched node so that EventHandler can later use it.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _scrollViewInternal]):
* Source/WebKit/UIProcess/ios/WKVelocityTrackingScrollView.h: Removed.
* Source/WebKit/UIProcess/ios/WKVelocityTrackingScrollView.mm: Removed.

Remove WKVelocityTrackingScrollView in favor of using WKBaseScrollView directly so that all scroll views can track velocities.

* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(ScrollingDeltaWindow::update):
(ScrollingDeltaWindow::reset):
(ScrollingDeltaWindow::averageVelocity const):
(-[WKBaseScrollView updateInteractiveScrollVelocity]):
(-[WKBaseScrollView interactiveScrollVelocityInPointsPerSecond]):

Give WKBaseScrollView the functionality that WKVelocityTrackingScrollView had.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(-[WKScrollingNodeScrollViewDelegate scrollViewDidScroll:]):

For sub-scrollable regions, call `updateInteractiveScrollVelocity` just like the main scroll view.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):

Since the scroll view that the scrolling animator uses is now dynamic, don't provide a scroll view when creating the animator;
instead, provide it on-demand.

(-[WKContentView _interpretKeyEvent:withContext:]):

Compute the scroll view that should be used for the keyboard scrolling animator, based on whichever view was focused/clicked/touched last.
Fallback to the main scroll view if no others are found.

(-[WKContentView keyboardScrollViewAnimatorDidFinishScrolling:]):

Use the correct scroll view when passing it to `_didFinishScrolling`.

* Source/WebKit/UIProcess/ios/WKKeyboardScrollingAnimator.h:
* Source/WebKit/UIProcess/ios/WKKeyboardScrollingAnimator.mm:
(-[WKKeyboardScrollViewAnimator init]):
(-[WKKeyboardScrollViewAnimator initWithScrollView:]): Deleted.

Instead of providing a scroll view at the beginning, it is provided on-demand.

(-[WKKeyboardScrollViewAnimator beginWithEvent:scrollView:]):
(-[WKKeyboardScrollViewAnimator beginWithEvent:]): Deleted.

Amend this function to allow specifying the current scroll view that should be targeted.
If there is no ongoing scroll, set the ivar _scrollView to the parameter's value.
If there is an ongoing scroll on a different scroll view, ignore the request on the new scroll view.

(-[WKKeyboardScrollViewAnimator scrollView]):

Allow querying for the current scroll view that the animator is using.

(-[WKKeyboardScrollViewAnimator didFinishScrolling]):

Set _scrollView to nil to indicate that there is now no scroll view that is currently keyboard scrolling.

* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView web_handleKeyEvent:]):
(-[WKPDFView web_initWithFrame:webView:mimeType:]):

Update callsites to use the new functions.

* Source/WebKit/UIProcess/ios/WKScrollView.h:

Change the base class.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Add a new member variable to keep track of the most recently tapped node.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleEditingKeyboardEvent):
(WebKit::WebPage::commitPotentialTap):

If there is no click node, then instead do a hit-test to find the relevant node that can be tapped, and update the last touched node.

Canonical link: https://commits.webkit.org/292342@main
